File: 2015/python/15.py
# Advent of Code 2015, day 15
# (c) aichingert

def prod(a):
    ans = 1
    for i in range(len(a)-1):
        if a[i] < 0:
            a[i] = 0
        ans *= a[i]
    return ans

def solve(ing,perms,part):
    ans = 0
    for i in range(len(perms)):
        sto = [0 for _ in range(len(ing[0]))]
        for j in range(len(ing)):
            for k in range(len(ing[j])):
                sto[k] += ing[j][k] * perms[i][j]

        if part and sto[-1] != 500:
            continue
        ans = max(ans,prod(sto))
    return ans
        
ing = []
perms = []

for l in open("../input/15").read().split('\n'):
    vls = l.split(' ')
    ing.append((int(vls[2][0:-1]), int(vls[4][0:-1]), int(vls[6][0:-1]), int(vls[8][0:-1]), int(vls[10])))

for i in range(101):
    for j in range(101):
        if i + j > 100: break
        for k in range(101):
            if i + j + k > 100: break
            for l in range(101):
                if i + j + k + l > 100: break
                if i + j + k +l == 100: perms.append((i,j,k,l))

print("Part 1:", solve(ing,perms,False))
print("Part 2:", solve(ing,perms,True))

File: 2015-haskell/rust/15.rs
// Advent of Code 2015, day 15
// (c) aichingert

fn solve(ing: &Vec<Vec<i32>>, perms: &Vec<Vec<i32>>, part: bool) -> i32 {
    let mut ans = 0;

    for perm in perms {
        let mut cur = vec![0; ing[0].len()];
        for i in 0..ing.len() {
            for j in 0..ing[i].len() {
                cur[j] += ing[i][j] * perm[i];
            }
        }

        let cal = cur.pop().unwrap();
        if !part && cal != 500 {
            continue;
        }

        ans = ans.max(
            cur.iter()
                .map(|v| if *v > 0 { *v } else { 0 })
                .fold(1, |sum, val| sum * val),
        );
    }

    ans
}

fn parse() -> (Vec<Vec<i32>>, Vec<Vec<i32>>) {
    let mut ing = Vec::new();
    let mut perms = Vec::new();

    for l in std::fs::read_to_string("../input/15").unwrap().lines() {
        let vls = l.split(' ').collect::<Vec<&str>>();
        ing.push(vec![
            vls[2][..vls[2].len() - 1].parse().unwrap(),
            vls[4][..vls[4].len() - 1].parse().unwrap(),
            vls[6][..vls[6].len() - 1].parse().unwrap(),
            vls[8][..vls[8].len() - 1].parse().unwrap(),
            vls[10].parse().unwrap(),
        ]);
    }

    for i in 1..100 {
        for j in 1..100 {
            if i + j > 100 {
                break;
            }
            for k in 1..100 {
                if i + j + k > 100 {
                    break;
                }
                for l in 1..100 {
                    if i + j + k + l > 100 {
                        break;
                    }
                    if i + j + k + l == 100 {
                        perms.push(vec![i, j, k, l]);
                    }
                }
            }
        }
    }

    (ing, perms)
}

fn main() {
    let (ing, perms) = parse();

    println!("Part 1: {}", solve(&ing, &perms, true));
    println!("Part 2: {}", solve(&ing, &perms, false));
}
